## Giá trị bị khuyết (Missing value)

Giá trị bị khuyết hay dữ liệu bị khuyết xảy ra khi __không có dữ liệu__ / __không có giá trị__ được lưu trữ cho các quan sát nhất định trong một biến.

Dữ liệu không hoàn thiện là vấn đề không thể tránh trong hầu hết các nguồn dữ liệu và có thể có tác động đáng kể đến các kết luận được rút ra từ dữ liệu.

### Tại sao dữ liệu bị khuyết?

Nguồn dữ liệu bị khuyết rất khác nhau. Đây chỉ là vài ví dụ:

- Giá trị bị khuyết do nó bị quên, bị mất hoặc không được lưu trữ cẩn thận.
- Các giá trị không tồn tại với một số quan sát.
- Giá trị không rõ hoặc không xác định.

Trong nhiều tổ chức, thông tin được thu thập vào một biểu mẫu qua nói chuyện với khách hàng trên điện thoại, hoặc khách hàng điền vào biểu mẫu trực tuyến. Thông thường, người nhập dữ liệu không điền tất cả các trường trong biểu mẫu. Nhiều trường không bắt buộc nên có thể dẫn đến thiếu giá trị.

Các lý do bỏ sót thông tin khá đa dạng: có thể người đó không muốn tiết lộ một số thông tin, ví dụ như thu nhập, hoặc họ không biết câu trả lời hay câu trả lời không áp dụng cho một trường hợp nhất định, hoặc ngược lại, người trong tổ chức muốn tiết kiệm thời gian cho khách hàng nên bỏ qua những câu hỏi mà họ cho là không liên quan.

Có các trường hợp khác mà giá trị của một biến nhất định không tồn tại. Ví dụ, trong biến 'tổng nợ tính theo phần trăm tổng thu nhập' (rất phổ biến trong dữ liệu tài chính), nếu một người không có thu nhập, thì tổng tỷ lệ phần trăm bằng 0 không tồn tại và do đó nó sẽ là một giá trị bị khuyết.

Điều quan trọng là phải hiểu **cách đưa dữ liệu bị khuyết vào tập dữ liệu**, nghĩa là **các cơ chế** đưa thông tin bị khuyết vào tập dữ liệu. Tùy thuộc vào cơ chế, chúng ta có thể chọn các cách xử lý giá trị bị khuyết khác nhau. Ngoài ra, nếu biết nguồn dữ liệu bị khuyết, chúng ta cũng có thể tìm các biện pháp kiểm soát nguồn đó và giảm lượng thông tin bị khuyết trong quá trình thu thập dữ liệu.


### Cơ chế dữ liệu bị khuyết

Có 3 cơ chế dẫn đến khuyết dữ liệu, 2 trong số đó liên quan đến dữ liệu bị khuyết ngẫu nhiên hoặc gần như ngẫu nhiên, và cơ chế thứ ba liên quan đến việc mất dữ liệu có hệ thống.

#### Khuyết hoàn toàn ngẫu nhiên (MCAR):

Một biến bị khuyết hoàn toàn ngẫu nhiên (MCAR) nếu xác suất bị khuyết là như nhau cho tất cả các quan sát. Khi dữ liệu là MCAR, hoàn toàn không có mối quan hệ nào giữa dữ liệu bị khuyết và bất kỳ giá trị nào khác được quan sát hoặc bị khuyết trong tập dữ liệu. Nói cách khác, những điểm dữ liệu bị khuyết đó là một tập hợp con ngẫu nhiên của dữ liệu. Những dữ liệu không có hệ thống khả năng sẽ bị khuyết cao hơn những dữ liệu khác. Nếu các giá trị cho các quan sát bị khuyết hoàn toàn ngẫu nhiên thì việc bỏ qua các trường hợp đó sẽ không làm sai lệch các suy luận được đưa ra.


#### Khuyết ngẫu nhiên (MAR):

MAR xảy ra khi có mối quan hệ giữa xu hướng của các giá trị bị khuyết và dữ liệu quan sát được. Nói cách khác, xác suất một quan sát bị khuyết phụ thuộc vào thông tin có sẵn (tức là các biến khác trong tập dữ liệu). Ví dụ, nếu nam giới có nhiều khả năng tiết lộ cân nặng của họ hơn nữ giới thì cân nặng là MAR. Thông tin về cân nặng sẽ bị khuyết ngẫu nhiên đối với những nam giới và nữ giới không tiết lộ cân nặng của họ, nhưng vì nam giới thường tiết lộ thông tin đó hơn nên sẽ có nhiều giá trị bị khuyết đối với nữ hơn.

Trong tình huống trên, nếu quyết định tiếp tục với biến có giá trị bị khuyết (trong trường hợp này là trọng số), chúng ta có thể hưởng lợi từ việc bao hàm giới tính để kiểm soát độ chệch về trọng lượng với các quan sát bị khuyết.


#### Khuyết không ngẫu nhiên (MNAR):

Dữ liệu bị khuyết không ngẫu nhiên (MNAR) khi có một cơ chế hoặc lý do cho việc đưa các giá trị bị khuyết vào tập dữ liệu. Ví dụ, MNAR sẽ xảy ra nếu mọi người không điền vào khảo sát trầm cảm vì mức độ trầm cảm của họ. Ở đây, việc khuyết dữ liệu có liên quan đến kết quả: trầm cảm. Tương tự, khi một công ty tài chính yêu cầu khách hàng cung cấp thông tin ngân hàng và danh tính để ngăn chặn gian lận danh tính, thường những kẻ lừa đảo mạo danh người khác sẽ không tải tài liệu lên vì chúng không có, chúng là lừa đảo. Do đó, có một mối quan hệ hệ thống giữa các tài liệu bị khuyết và mục tiêu mà chúng ta muốn dự đoán: gian lận.

Hiểu cơ chế dữ liệu bị khuyết rất quan trọng khi quyết định sử dụng phương pháp nào để tính các giá trị bị khuyết. 

====================================================================================================

## Trong mô phỏng này:

Trong các cell sau, chúng ta sẽ:

- Học cách phát hiện và xác định các giá trị bị khuyết.

- Thử xác định 3 cơ chế đề xuất dữ liệu bị khuyết khác nhau.

Chúng ta sẽ sử dụng tập dữ liệu toy Loan và titanic.


In [1]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt

# hiển thị tổng số cột có trong tập dữ liệu
pd.set_option('display.max_columns', None)

In [4]:
## Yêu cầu 1: load tập dữ liệu titanic
## VIẾT CODE Ở ĐÂY:
data = pd.read_csv('./dataset/titanic.csv')

<details><summary>Gợi ý</summary>


Sử dụng hàm **read_csv** của thư viện **pandas**.

```

</details>


In [5]:
## Yêu cầu 2: Kiểm tra 5 hàng đầu tiên
## VIẾT CODE Ở ĐÂY:
data.head()

pclass  survived                                             name     sex  \
0       1         1                    Allen, Miss. Elisabeth Walton  female   
1       1         1                   Allison, Master. Hudson Trevor    male   
2       1         0                     Allison, Miss. Helen Loraine  female   
3       1         0             Allison, Mr. Hudson Joshua Creighton    male   
4       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   

       age  sibsp  parch  ticket      fare cabin embarked boat   body  \
0  29.0000      0      0   24160  211.3375    B5        S    2    NaN   
1   0.9167      1      2  113781  151.5500   C22        S   11    NaN   
2   2.0000      1      2  113781  151.5500   C22        S  NaN    NaN   
3  30.0000      1      2  113781  151.5500   C22        S  NaN  135.0   
4  25.0000      1      2  113781  151.5500   C22        S  NaN    NaN   

                         home.dest  
0                     St Louis, MO  
1  Montreal, PQ / Chesterville, ON  
2  Montreal, PQ / Chesterville, ON  
3  Montreal, PQ / Chesterville, ON  
4  Montreal, PQ / Chesterville, ON

<details><summary>Gợi ý</summary>


Sử dụng phương thức **head**.



</details>


Trong python, giá trị bị khuyết được lưu thành NaN, xem ví dụ hàng đầu với biến cabin.

In [12]:
## Yêu cầu 3: xác định tổng số giá trị bị khuyết
## VIẾT CODE Ở ĐÂY:
data.isnull().sum()

pclass          0
survived        0
name            0
sex             0
age           263
sibsp           0
parch           0
ticket          0
fare            1
cabin        1014
embarked        2
boat          823
body         1188
home.dest     564
dtype: int64

<details><summary>Gợi ý</summary>

Sử dụng phương thức **isnull** kết hợp với phương thức **sum**.

</details>


Có 263 giá trị bị khuyết cho age, 1014 cho cabin và 2 cho embarked.

In [13]:
## Yêu cầu 4: sử dụng phương thức mean sau isnull để hiển thị số phần trăm giá trị bị khuyết cho từng biến
## VIẾT CODE Ở ĐÂY:
data.isnull().mean()

pclass       0.000000
survived     0.000000
name         0.000000
sex          0.000000
age          0.200917
sibsp        0.000000
parch        0.000000
ticket       0.000000
fare         0.000764
cabin        0.774637
embarked     0.001528
boat         0.628724
body         0.907563
home.dest    0.430863
dtype: float64

Dữ liệu bị khuyết trong biến: age (20%), cabin (77%) và embarked ( ~0.2%).

## Cơ chế khuyết dữ liệu

### Khuyết không ngẫu nhiên (MNAR): Các giá trị bị khuyết có hệ thống

Trong tập dữ liệu Titanic, giá trị bị khuyết của các biến **age** và **cabin** đều được giới thiệu một cách có hệ thống. Không thể thiết lập **age** hoặc **cabin** với những người không sống sót. Ngược lại những người sống sót có thể được yêu cầu cung cấp thông tin đó.

Chúng ta có thể suy ra điều này bằng cách xem dữ liệu không?

Trong tình huống như vậy, chúng ta có thể dư kiến số lượng các giá trị bị khuyết lớn hơn cho những người không sống sót.

Hãy xem.

In [17]:
## Yêu cầu 5: Hãy tạo một biến nhị phân chỉ ra giá trị cabin có bị khuyết hay không 
## VIẾT CODE Ở ĐÂY:
data['cabin_null'] = np.where(data['cabin'].isnull(), 1, 0)

In [27]:
# hãy đánh giá tỷ lệ giá trị bị khuyết trong
# cabin cho những người sống với không sống sót
data.cabin_null.mean()*100
# biến Survived nhận giá trị 1 nếu hành khách 
# sống sót, nếu không là 0

## Yêu cầu 6:
# nhóm dữ liệu theo Survived với Non-Survived
# và tìm tỷ lệ null cho cabin
## VIẾT CODE Ở ĐÂY:
data.groupby(['survived'])['cabin_null'].mean()

survived
0    0.873918
1    0.614000
Name: cabin_null, dtype: float64

In [ ]:
# một cách khác để thực hiện điều trên với ít dòng hơn
# code :)

data['cabin'].isnull().groupby(data['survived']).mean()

survived
0    0.873918
1    0.614000
Name: cabin, dtype: float64

Chúng ta thấy rằng tỷ lệ giá trị bị khuyết của những người không sống sót (87%) cao hơn những người sống sót (60%). Phát hiện này phù hợp với giả thuyết rằng dữ liệu bị khuyết vì không thể truy xuất thông tin sau khi chết.

**Lưu ý**: Do đó, để thực sự củng cố liệu dữ liệu MNAR, chúng ta cần làm quen với cách thu thập dữ liệu. Phân tích tập dữ liệu chỉ có thể chỉ chúng ta đi đúng hướng hoặc giúp xây dựng các giả định. 

In [28]:
## Yêu cầu 7:
# Hãy thực hiện tương tự với biến age:

# rước tiên tạo biến nhị phân để chỉ
# giá trị của age có bị khuyết không
## VIẾT CODE Ở ĐÂY:
data['age_null'] = np.where(data['age'].isnull(), 1, 0)

# và xem giá trị trung bình ở các nhóm sống sót khác nhau:
## VIẾT CODE Ở ĐÂY:
data.groupby(['survived'])['age_null'].mean()

survived
0    0.234858
1    0.146000
Name: age_null, dtype: float64

In [29]:
# hoặc tương tự với code đơn giản hơn :)

data['age'].isnull().groupby(data['survived']).mean()

survived
0    0.234858
1    0.146000
Name: age, dtype: float64

Một lần nữa, chúng ta quan sát thấy số lượng dữ liệu bị khuyết cao hơn với những người không sống sót sau thảm kịch. Do đó, phân tích cho thấy có một sự mất mát dữ liệu có hệ thống: những người không sống sót có xu hướng khuyết nhiều thông tin hơn. Cũng có thể do phương pháp thu thập thông tin góp phần tạo ra những dữ liệu bị khuyết này.

### Khuyết dữ liệu hoàn toàn ngẫu nhiên (MCAR)

In [30]:
# Trong tập dữ liệu titanic cũng có giá trị bị khuyết
# cho biến Embarked.
# Hãy quan sát.

## Yêu cầu 8:
# Cắt dataframe để chỉ hiển thị các quan sát
# với các giá trị bị khuyết cho Embarked

data[data['embarked'].isnull()]

pclass  survived                                       name     sex  \
168       1         1                        Icard, Miss. Amelie  female   
284       1         1  Stone, Mrs. George Nelson (Martha Evelyn)  female   

      age  sibsp  parch  ticket  fare cabin embarked boat  body  \
168  38.0      0      0  113572  80.0   B28      NaN    6   NaN   
284  62.0      0      0  113572  80.0   B28      NaN    6   NaN   

          home.dest  cabin_null  age_null  
168             NaN           0         0  
284  Cincinatti, OH           0         0

Icard là người giúp việc của bà Stone, 2 người phụ nữ này đã đi du lịch cùng nhau.

Dường như không có dấu hiệu cho thấy thông tin bị khuyết trong biến Embarked phụ thuộc vào bất kỳ biến nào khác; thực tế, những phụ nữ này sống sót, nghĩa là họ có thể đã được hỏi về thông tin này.

Rất có thể các giá trị đã bị mất tại thời điểm xây dựng tập dữ liệu.

Nếu các giá trị này là MCAR thì xác suất khuyết dữ liệu của 2 người phụ nữ này bằng với xác suất khuyết các giá trị với bất kỳ người nào khác trên Titanic. Tất nhiên điều này sẽ khó để chứng minh. Hy vọng đây là một minh chứng. 

### Khuyết dữ liệu ngẫu nhiên (MAR)

Đối với ví dụ này, chúng ta sẽ sử dụng tập dữ liệu loan book toy từ công ty cho vay ngang hàng giả tưởng này.

Chúng ta sẽ xem xét các biến employment và years in employment, cả hai đều được người vay khai báo tại thời điểm đăng ký vay.

Trong ví dụ này, dữ liệu bị khuyết trong employment có liên quan đến dữ liệu bị khuyết trong time_employed. 

In [35]:
## Yêu cầu 9:
# load các cột cho bài tập này 

## VIẾT CODE Ở ĐÂY:
data = pd.read_csv('./dataset/loan.csv', usecols=['employment','time_employed'])

data.head()

employment time_employed
0       Teacher     <=5 years
1    Accountant     <=5 years
2  Statistician     <=5 years
3         Other     <=5 years
4    Bus driver      >5 years

In [36]:
## Yêu cầu 10:
# kiểm tra tỷ lệ phần trăm dữ liệu bị khuyết
## VIẾT CODE Ở ĐÂY:
data.isnull().mean()

employment       0.0611
time_employed    0.0529
dtype: float64

Chúng ta thấy cả 2 biến có tỷ lệ quan sát bị khuyết như nhau.

In [49]:
## Yêu cầu 11:
# kiểm tra các kiểu công việc khác nhau

# số công việc khác nhau
## VIẾT CODE Ở ĐÂY:
print('Number of employments: {}'.format(
    len(data['employment'].unique())))

# Ví dụ về công việc
## VIẾT CODE Ở ĐÂY:
data['employment'].unique()

Number of employments: 12


array(['Teacher', 'Accountant', 'Statistician', 'Other', 'Bus driver',
       'Secretary', 'Software developer', 'Nurse', 'Taxi driver', nan,
       'Civil Servant', 'Dentist'], dtype=object)

<details><summary> Gợi ý: </summary>

Sử dụng hàm `unique()` 
</details>

Chúng ta thấy thông tin bị khuyết (NaN) và vài nghề nghiệp khác.

In [50]:
## Yêu cầu 12:
#  kiểm tra biến time_employed
## VIẾT CODE Ở ĐÂY:
data['time_employed'].unique()

array(['<=5 years', '>5 years', nan], dtype=object)

<details><summary> Gợi ý: </summary>

Sử dụng hàm `unique()` 
</details>

Khách hàng không thể nhập giá trị cho employment time nếu họ không làm việc. Họ có thể là sinh viên, đã nghỉ hưu, tự kinh doanh hoặc làm việc ở nhà. Nhưng hãy xem 2 biến này liên hệ với nhau như thế nào.

In [52]:
## Yêu cầu 13:
# tính tỷ lệ dữ liệu bị khuyết cho biến time_employed
# cho những khách khai báo công việc

# những khách khai báo công việc
## VIẾT CODE Ở ĐÂY:
t = data[~data['employment'].isnull()]

# phần trăm dữ liệu bị khuyết trong time_employed
## VIẾT CODE Ở ĐÂY:
t['time_employed'].isnull().mean()

0.0005325380764724678

In [53]:
## Yêu cầu 14:
# thực hiện tương tự với những người vay mà không
# khai báo công việc


# các khách không khai báo công việc
## VIẾT CODE Ở ĐÂY:
t = data[data['employment'].isnull()]

# phần trăm dữ liệu bị khuyết trong time_employed
## VIẾT CODE Ở ĐÂY:
t['time_employed'].isnull().mean()

0.8576104746317512

Số lượng người vay đã khai báo nghề nghiệp và khuyết giá trị trong time_employed là tối thiểu trong khi những khách hàng không khai báo nghề nghiệp hoặc việc làm, hầu hết đang báo cáo thiếu giá trị bị thiếu trong biến time_employed,

Điều này hỗ trợ thêm rằng các giá trị bị khuyết trong employment có liên quan đến các giá trị bị khuyết trong time_employed.

Đây là ví dụ về MAR. 